### PR TITLE
scaletempo2: add intrinsics for aarch64

### PR DESCRIPTION
**Pull request by CounterPillow**
Loads four samples at once, probably still room for improvement ~~(only about 2x speedup vs. -O3)~~(almost 3x speedup with real world parameters). Correct to at most an epsilon of 0.0001 (didn't try any lower.)

### CODE REVIEW

**Comment by sfan5** [comment]
I think by the (unused) `--enable-asm` switch in waf it's implied that inline assembly in mpv is supposed to work somewhat like ffmpeg. Specifically, reside in a separate file.
No idea if that transfers nicely to intrinsics (ffmpeg has only(?) fully hand-rolled asm).

File: audio/filter/af_scaletempo2_internals.c [hunk 1]
```diff
@@ -115,15 +119,36 @@ static void multi_channel_dot_product(
...
+            dp = vfmaq_f32(dp, a_ld, b_ld);
+        }
+        float after_simd = vaddvq_f32(dp);
+        for(; n < num_frames; n++) {
+            after_simd += *ch_a++ * *ch_b++;
+        }
+        dot_product[k] = after_simd;
+    }
+#else
+    for (; k < channels; ++k) {
```
Comment (sfan5) [inline audio/filter/af_scaletempo2_internals.c:142]:
```suggestion
    for (int k = 0; k < channels; ++k) {
```

File: audio/filter/af_scaletempo2_internals.c [hunk 2]
```diff
@@ -115,15 +119,36 @@ static void multi_channel_dot_product(
...
         const float* ch_a = a[k] + frame_offset_a;
         const float* ch_b = b[k] + frame_offset_b;
+        float dp = 0;
         for (int n = 0; n < num_frames; ++n) {
-            dot_product[k] += *ch_a++ * *ch_b++;
+            dp += *ch_a++ * *ch_b++;
```
Comment (sfan5) [inline audio/filter/af_scaletempo2_internals.c:147]: also this change should still be a separate commit

**Comment by haasn** [comment]
Out of curiosity, can you compare this against https://github.com/mpv-player/mpv/pull/8853 ?

I think the advantage of my code is that it automatically takes advantage of whatever is available on the target platform, and it seems to perform just as well (even better) than hand-rolled AVX2 for me.

**Comment by CounterPillow** [comment]
> Out of curiosity, can you compare this against #8853 ?
> 
> I think the advantage of my code is that it automatically takes advantage of whatever is available on the target platform, and it seems to perform just as well (even better) than hand-rolled AVX2 for me.

Your speedup is between 1.02x and 1.25x on my in-order ROCK64, so mine's faster at between 1.47x to 2.5x depending on input size (somehow it's less with big inputs?) I compiled with `-march=native -O2` and also -O3, not sure what else is needed to make gcc do its thing here for sure.

Though my target is a bit exotic so your method is probably fine.

>  No idea if that transfers nicely to intrinsics (ffmpeg has only(?) fully hand-rolled asm).

No point to do this for intrinsics. No assembler is required. Even if I wrote this in hand-rolled AArch64 assembly, it's be using GAS, in which case anyone with clang or gcc would be able to build it, so the switch doesn't make much sense outside of, idk, MSVC on ARM? But they support intrinsics I think so yeah.

Not gonna touch waf if I can find enough excuses not to do so

File: audio/filter/af_scaletempo2_internals.c (hunk 2 above)
Comment (CounterPillow) [reply audio/filter/af_scaletempo2_internals.c:147]: I'm just gonna do it the original way here because the compiler seems to optimise it to the same thing and everyone loves memset for some reason

**Comment by sfan5** [comment]
> No point to do this for intrinsics. No assembler is required.

**Comment by CounterPillow** [comment]
> > No point to do this for intrinsics. No assembler is required.
> 
> Should this option even apply to it, then?
> 
> https://github.com/mpv-player/mpv/blob/6c1dd02f322eba2990f4cd9a6b8ddd7a94fbf1dc/wscript#L115-L118

Inline assembly is different from intrinsics.

Inline assembly uses special syntax to make the compiler process some assembler code that's thrown into the regular C code, with registers bound to certain variables. GCC supports some version of it, Clang supports an older version of GCC's, and MSVC does its own thing as MSVC does IIRC.

Intrinsics on the other hand are implementations provided by the compiler. It's a header file you can include and get the same functions on all compilers, as far as I understand, just with compiler-specific implementations (which behave exactly the same usually)

File: audio/filter/af_scaletempo2_internals.c (hunk 2 above)
Comment (sfan5) [reply audio/filter/af_scaletempo2_internals.c]: It doesn't tho. Just that it's apparently as fast.

**Comment by sfan5** [comment]
Yes I know. Just wondering if there should be a way to disable this implementation.

**Comment by CounterPillow** [comment]
Okay so I did some comparisons between mine and haasn's, and while haasn's works across architectures, mine is about 1.44 times faster than his.

One option would be to just throw my version away; aarch64 isn't a huge target, and haasn's code does result in a significant enough speedup. The other option would be to merge both versions, so that mine is chosen on aarch64, and his on other architectures.

As for whether we ever want to disable the AArch64 implementation, I don't know. If it ever results in a bug, somebody should report it so that it can be fixed. That includes build failures; I don't even know what aarch64 compiler that currently compiles all of mpv would choke on intrinsics. Once I get an AArch64 SBC with different cores, I'll repeat the experiment to see if I still am ahead; if not, we can throw my code out at that point.

**Comment by CounterPillow** [comment]
Update: with clang, haasn is faster than me, but we're both faster than gcc.

**Comment by CounterPillow** [comment]
Closing this because it's basically made obsolete by haasn's stuff if gcc ever gits gud.
